File: scripts/RollershutterPublisher.py
#!/usr/bin/python
import rospy
from openhab_msgs.msg import RollershutterCommand

class RollershutterPublisher(object):
    """Node example class."""

    def __init__(self, item_name):

        self.item_name = item_name
        self.pub = rospy.Publisher("/openhab/items/" + self.item_name + "/command", RollershutterCommand, queue_size=10)
        self.rate = rospy.Rate(10) # 10hz

        # Initialize message variables.
        self.enable = False
        self.message = None

        if self.enable:
            self.start()
        else:
            self.stop()

    def start(self):
        """Turn on publisher."""
        self.enable = True
        self.pub = rospy.Publisher("/openhab/items/" + self.item_name + "/command", RollershutterCommand, queue_size=10)

        while not rospy.is_shutdown():

            self.message = RollershutterCommand()

            self.message.command = "DOWN"
            self.message.iscommand = True
            self.message.ispercentage = False
            self.message.percentage = 0
            
            self.message.isnull = False

            self.message.header.stamp = rospy.Time.now()
            self.message.header.frame_id = "/base_link"
            self.message.item = self.item_name

            message = "Publishing to %s at %s: command = %s, iscommand = %s, ispercentage = %s, percentage = %s" % (self.message.item, rospy.get_time(), self.message.command, self.message.iscommand, self.message.ispercentage, self.message.percentage)
            rospy.loginfo(message)

            self.pub.publish(self.message)
            self.rate.sleep()

    def stop(self):
        """Turn off publisher."""
        self.enable = False
        self.pub.unregister()

# Main function.
if __name__ == "__main__":
    # Initialize the node and name it.
    rospy.init_node("RollershutterPublisherNode", anonymous=True)
    # Go to class functions that do all the heavy lifting.

    rollershutterPublisher = RollershutterPublisher("testRollershutter")

    try:
        rollershutterPublisher.start()
    except rospy.ROSInterruptException:
        pass
    # Allow ROS to go to all callbacks.
    # spin() simply keeps python from exiting until this node is stopped
    rospy.spin()

File: src/RollershutterPublisher.cpp
#include <ros/ros.h>
#include <openhab_msgs/RollershutterCommand.h>

class RollershutterPublisher {
  public:
    RollershutterPublisher(std::string item_name) {
        item_name_ = item_name;
        pub_ = nh_.advertise<openhab_msgs::RollershutterCommand>("/openhab/items/" + item_name_ + "/command", 10);
        rate_ = ros::Rate(10); // 10hz
    }

    void start() {
        enable_ = true;
        pub_ = nh_.advertise<openhab_msgs::RollershutterCommand>("/openhab/items/" + item_name_ + "/command", 10);

        while (ros::ok()) {
            message_.command = "DOWN";
            message_.iscommand = true;
            message_.ispercentage = false;
            message_.percentage = 0;
            message_.isnull = false;
            message_.header.stamp = ros::Time::now();
            message_.header.frame_id = "/base_link";
            message_.item = item_name_;

            std::stringstream ss;
            ss << "Publishing to " << message_.item << " at " << ros::Time::now() << ": command = " << message_.command << ", iscommand = " << message_.iscommand << ", ispercentage = " << message_.ispercentage << ", percentage = " << message_.percentage;
            ROS_INFO_STREAM(ss.str());

            pub_.publish(message_);
            rate_.sleep();
        }
    }

    void stop() {
        enable_ = false;
        pub_.shutdown();
    }

  private:
    ros::NodeHandle nh_;
    ros::Publisher pub_;
    ros::Rate rate_;
    std::string item_name_;
    bool enable_ = false;
    openhab_msgs::RollershutterCommand message_;
};

int main(int argc, char **argv) {
    ros::init(argc, argv, "RollershutterPublisherNode");
    RollershutterPublisher rollershutterPublisher("testRollershutter");

    try {
        rollershutterPublisher.start();
    } catch (const ros::Exception& e) {
        ROS_ERROR_STREAM("RollershutterPublisherNode exception: " << e.what());
    }

    ros::spin();
    return 0;
}
